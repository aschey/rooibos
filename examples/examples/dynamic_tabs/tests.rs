use rooibos::tester::TestHarness;

use crate::app;

macro_rules! assert_snapshot {
    ($harness:expr) => {
        let buffer = $harness.buffer().await;
        insta::with_settings!({
            snapshot_path => "./snapshots"
        }, {
            insta::assert_debug_snapshot!(buffer);
        });
    };
}

#[rooibos::test]
async fn test_tabs() {
    let mut harness = TestHarness::new(40, 10).await;
    harness.mount(app).await;

    assert_snapshot!(harness);

    let tab2_pos = harness.get_position_of_text("Tab2").await;
    harness.click_pos(tab2_pos);
    harness
        .wait_for(async |harness, _| harness.terminal_view().await.contains("tab2"))
        .await
        .unwrap();
    assert_snapshot!(harness);

    let add_pos = harness.get_position_of_text("+").await;
    harness.click_pos(add_pos);
    harness
        .wait_for(async |harness, _| harness.terminal_view().await.contains("Tab3"))
        .await
        .unwrap();
    assert_snapshot!(harness);

    let tab3_pos = harness.get_position_of_text("Tab3").await;
    harness.click_pos(tab3_pos);
    harness
        .wait_for(async |harness, _| harness.terminal_view().await.contains("tab3"))
        .await
        .unwrap();

    let tab2_close = harness.get_nth_position_of_text("âœ•", 2).await;
    harness.click_pos(tab2_close);
    harness
        .wait_for(async |harness, _| harness.terminal_view().await.contains("tab2"))
        .await
        .unwrap();
    assert_snapshot!(harness);

    harness.exit().await;
}
