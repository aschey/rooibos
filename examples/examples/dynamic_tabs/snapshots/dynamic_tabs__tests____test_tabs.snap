---
source: examples/examples/dynamic_tabs/tests.rs
expression: buffer
---
Buffer {
    area: Rect { x: 0, y: 0, width: 40, height: 10 },
    content: [
        "                                        ",
        " ┌Demo───────────────┐╭───╮             ",
        " │ Tab1  ✕ │ Tab2  ✕ ││ + │             ",
        " └───────────────────┘╰───╯             ",
        "  tab1                                  ",
        "                                        ",
        "                                        ",
        "                                        ",
        "                                        ",
        "                                        ",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Reset, modifier: NONE,
        x: 22, y: 1, fg: Gray, bg: Reset, modifier: NONE,
        x: 27, y: 1, fg: Reset, bg: Reset, modifier: NONE,
        x: 3, y: 2, fg: Yellow, bg: Reset, modifier: NONE,
        x: 7, y: 2, fg: Reset, bg: Reset, modifier: NONE,
        x: 9, y: 2, fg: Red, bg: Reset, modifier: NONE,
        x: 10, y: 2, fg: Reset, bg: Reset, modifier: NONE,
        x: 19, y: 2, fg: Red, bg: Reset, modifier: NONE,
        x: 20, y: 2, fg: Reset, bg: Reset, modifier: NONE,
        x: 22, y: 2, fg: Gray, bg: Reset, modifier: NONE,
        x: 23, y: 2, fg: Reset, bg: Reset, modifier: NONE,
        x: 24, y: 2, fg: Green, bg: Reset, modifier: NONE,
        x: 25, y: 2, fg: Reset, bg: Reset, modifier: NONE,
        x: 26, y: 2, fg: Gray, bg: Reset, modifier: NONE,
        x: 27, y: 2, fg: Reset, bg: Reset, modifier: NONE,
        x: 22, y: 3, fg: Gray, bg: Reset, modifier: NONE,
        x: 27, y: 3, fg: Reset, bg: Reset, modifier: NONE,
    ]
}
